### PR TITLE
Fix error in filtering documentation

### PR DESCRIPTION
**Pull request by cbrnr**
There is an error in the transfer function equation (a 1 term is missing in the denominator). I suggest to remove the second version with sum signs altogether, because the first version contains all information and is just as easy to read.

Screen Shot 2022-03-25 at 08 48 40

### CODE REVIEW

**Comment by agramfort** [comment]
can you share the new rendered doc when you have it? 🙏

**Comment by cbrnr** [comment]
@agramfort sure, will do! I've included a screenshot of the equation with my proposed changes!

**Comment by agramfort** [comment]
good catch ! 

I think what got us confused is scipy https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.lfilter.html#scipy.signal.lfilter that still requires you to pass a[0] in the filtering parameters.

**Comment by cbrnr** [comment]
I think a[0] can be ≠ 1 for some IIR filters, so this would be a more general case. It is 1 for most IIR filters though (at least according to the [Wikipedia article](https://en.wikipedia.org/wiki/Infinite_impulse_response#Transfer_function_derivation)).

**Comment by cbrnr** [comment]
Rendered page: https://42836-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/preprocessing/25_background_filtering.html

My suggestion is to remove the second line instead of correcting it.

Also, I was trying to remove the "quoted" style of enumerated lists to save some horizontal space, but now the first lines are bold. Is there any documentation on RST and the style we're using so that this doesn't seem like arbitrarily changing Markdown syntax to me?

**Comment by cbrnr** [comment]
BTW, there used to be a check which directly linked to the built HTML documentation artifact – is this gone now?

**Comment by agramfort** [comment]
LGTM

link should be still there

+0 on removing the second line.

**Comment by cbrnr** [comment]
> link should be still there

I don't see it. There are only two checks related to the docs:
- ci/circleci: build_docs
- ci/circleci: linkcheck

**Comment by agramfort** [comment]
indeed I don't know why it's not available anymore. Maybe @larsoner has an idea

**Comment by larsoner** [comment]
> BTW, there used to be a check which directly linked to the built HTML documentation artifact – is this gone now?

Probably https://github.com/larsoner/circleci-artifacts-redirector-action/issues/19

> I was trying to remove the "quoted" style of enumerated lists to save some horizontal space, but now the first lines are bold.

This is probably an indentation issue, whereby you accidentally created a "definition list"

> Is there any documentation on RST and the style we're using so that this doesn't seem like arbitrarily changing Markdown syntax to me?

RST documentation is pretty sparse. When I hit these problems I usually use an online RST renderer, or VSCode's "preview" for the RST extension

File: tutorials/preprocessing/25_background_filtering.py [hunk 1]
```diff
@@ -66,16 +66,16 @@
...
 In other words, the output at time :math:`n` is determined by a sum over
 
-    1. the numerator coefficients :math:`b_k`, which get multiplied by
-       the previous input values :math:`x(n-k)`, and
-    2. the denominator coefficients :math:`a_k`, which get multiplied by
-       the previous output values :math:`y(n-k)`.
+1. the numerator coefficients :math:`b_k`, which get multiplied by the previous
+input values :math:`x(n-k)`, and
+2. the denominator coefficients :math:`a_k`, which get multiplied by the
+previous output values :math:`y(n-k)`.
```
Comment (larsoner) [inline tutorials/preprocessing/25_background_filtering.py:78]: Will probably fix it (to be tested)
```suggestion
1. the numerator coefficients :math:`b_k`, which get multiplied by the previous
   input values :math:`x(n-k)`, and
2. the denominator coefficients :math:`a_k`, which get multiplied by the
   previous output values :math:`y(n-k)`.
```

**Comment by cbrnr** [comment]
> Probably https://github.com/larsoner/circleci-artifacts-redirector-action/issues/19

Maybe, but the check doesn't even show up anymore (as opposed to containing an invalid URL).

> This is probably an indentation issue, whereby you accidentally created a "definition list"

True, but removing the indentation does not seem to create a list anymore:

Screen Shot 2022-03-27 at 09 09 35

https://42837-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/preprocessing/25_background_filtering.html

I think it's because there are line breaks, which I have removed now. Why is the first part in a (raw) string, whereas all the other text is contained in comments?

> RST documentation is pretty sparse. When I hit these problems I usually use an online RST renderer, or VSCode's "preview" for the RST extension

OK. Would it make sense to discuss switching to MyST?

**Comment by cbrnr** [comment]
Is the delay plot supposed to show e.g. 0.5s in this example? Then the scaling is way off and should be fixed (how?):
Screen Shot 2022-03-29 at 11 14 20

**Comment by cbrnr** [comment]
Latest render: https://output.circle-artifacts.com/output/job/8ea9fd5b-d6d0-4028-a788-62676839d0a3/artifacts/0/dev/auto_tutorials/preprocessing/25_background_filtering.html

**Comment by cbrnr** [comment]
Open questions:
- Where do I put the `# noqa` tags for a cell (of comments)? Putting it right at the end of a line doesn't work (because this gets rendered although the flake8 warnings are suppressed).
- I tried to update the defaults for the MNE versions, i.e. instead of "MNE 0.16" I wrote "MNE ≥0.16". I don't know if this is correct, but I think the reader will at least want to know what the *current* default is. Can anyone please check?
- Scaling on the y-axis in all Delay plots might be off.
- The summarizing figures (one right at the end of the FIR section, and the other in the "Applying IIR filters" section) are a bit misleading, because in the left column they show non-causal responses for all filters but the minimal-phase filter. I think we should include the delay for all filters here to show that the minimal-phase filter has the smallest delay.
- I removed references to MATLAB functions because I think these are not needed. I suspect most people have never used these MATLAB functions before.

**Comment by larsoner** [comment]
> Where do I put the # noqa tags for a cell (of comments)? Putting it right at the end of a line doesn't work (because this gets rendered although the flake8 warnings are suppressed).

There is no good way to do this that I know of. We could maybe use a file-level noqa of E501 (?) but I would rewrite the lines if possible. So far I think we have always found a way. I can try and push a commit if it would help.

> Is the delay plot supposed to show e.g. 0.5s in this example? Then the scaling is way off and should be fixed (how?):
... Scaling on the y-axis in all Delay plots might be off.

The one you reference above at least should have 0 delay (not 0.5 sec) because it's a non-causal, zero-phase filter. By definition the delay is zero everywhere, otherwise it is not a zero-phase filter.

> are a bit misleading, because in the left column they show non-causal responses for all filters except for the minimal-phase filter.

I don't think this is misleading, but perhaps could be clarified. Non-causal, zero-phase filters should show zero delay by definition. Perhaps you're thinking of a closely related *causal, linear-phase filter* which could be constructed using identical kernel values, but shifted such that its `h[n]` has non-zero values starting from `n=0` and continuing to `2*N-1` samples, rather than the zero-phase variant, which has non-zero values going from `-N` to `N` (and must be even around `n=0`). But such a linear-phase filter is a different from the zero-phase ones implemented and used in MNE-Python when you do `phase='zero'`, so showing the delay of that related linear phase filter would be misleading here.

> I think we should include the delay for all filters here to show that the minimal-phase filter has the smallest delay.

The minimal phase filter does indeed have less delay than a causal, linear phase filter that starts non-zero values at `n=0`. However, it does not have less delay than the zero-phase filter. The causal, linear phase filter is not available in MNE-Python's filtering functions, so it's not really emphasized here. (We could expose it easily with something like `phase='linear'`, but YAGNI so far at least.)

Maybe some of our wording could be improved to clarify this stuff?

**Comment by cbrnr** [comment]
> There is no good way to do this that I know of. We could maybe use a file-level noqa of E501 (?) but I would rewrite the lines if possible. So far I think we have always found a way. I can try and push a commit if it would help.

I can try to shorten, but this means I am restricted to 79 characters. I'll see if this is possible here.

> The one you reference above at least should have 0 delay (not 0.5 sec) because it's a non-causal, zero-phase filter. By definition the delay is zero everywhere, otherwise it is not a zero-phase filter.

Regarding delays, I think it is not clear that this tutorial talks about zero-phase filters. For example, in the "Designing FIR filters" section, it says: "...and the filter has a correspondingly much longer group delay (again equal to half the filter length, or 0.5 seconds)".

It is now clear that the delay plots always show zero, but it is confusing that the y-axis scaling is different for the various plots: the first one ranges from -50 to 50, the second one from -500 to 500, and the third one from -5000 to 5000.

Even in cases where there *is* a non-zero delay (the minimum phase filter), the scaling is so large that it basically looks like 0.

I suggest that we do not show the delay plots in this tutorial, because they add more confusion than they help (they basically show a straight line at zero, but the y-axis scaling differs from plot to plot).

Regarding the two summary plots, it was not clear to me that this shows non-causal filters. Given that it is a background tutorial on filtering, I think it would make sense to show causal filters as well as non-causal ones. However, a quick fix could be to explicitly state that all FIR filters shown in the figure are non-causal and therefore have zero delay.

I basically think of causal FIR filters as the starting point, and to make the zero-phase (non-causal), you shift the filter by a certain amount of samples.

In summary, I am not suggesting to add any new features to MNE (although causal FIR filters could be useful in some cases, but I agree that YAGNI until someone wants this). Instead, rewording the tutorial a little should make things clear.

So these are the issues I would like to address:

- Deal with `# noqa`
- Delay plots change scaling all the time, I suggest to remove them altogether
- Make it clear that the summary plots show non-causal FIR filters (except for the minimum phase one)

**Comment by larsoner** [comment]
I'd like to keep the delay plots in some form. It's useful to show that zero phase stays zero phase/delay, and that minimum phase has some delay. Let's change the ylims so that they make the plots more informative rather than remove them.

**Review by larsoner — APPROVED** [review]
@cbrnr feel free to merge if you're happy:

https://output.circle-artifacts.com/output/job/a79291e3-205a-4dd1-b4f5-7e6a03bb1c48/artifacts/0/dev/auto_tutorials/preprocessing/25_background_filtering.html

**Review by agramfort — APPROVED** [review]

**Comment by cbrnr** [comment]
Perfect, thanks @larsoner!
